### PR TITLE
Change all mailboxes to channels

### PR DESCRIPTION
**Pull request by TiraelSedai**
PR for issue: Change all mailboxes to channels to improve performance #168

### CODE REVIEW

File: src/Pulsar.Client/Internal/ConsumerImpl.fs [hunk 1]
```diff
@@ -32,26 +33,26 @@ type internal ConsumerMessage<'T> =
...
+    | Acknowledge of MessageId * AckType * Transaction option * TaskCompletionSource<Task<Unit>> option
     | NegativeAcknowledge of MessageId
-    | RedeliverUnacknowledged of RedeliverSet * AsyncReplyChannel<unit>
-    | RedeliverAllUnacknowledged of AsyncReplyChannel<unit>
-    | SeekAsync of SeekType * AsyncReplyChannel<ResultOrException<unit>>
-    | HasMessageAvailable of AsyncReplyChannel<Task<bool>>
+    | RedeliverUnacknowledged of RedeliverSet * TaskCompletionSource<unit>
+    | RedeliverAllUnacknowledged of TaskCompletionSource<unit>
+    | SeekAsync of SeekType * TaskCompletionSource<ResultOrException<unit>>
+    | HasMessageAvailable of TaskCompletionSource<Task<bool>>
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConsumerImpl.fs]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/ConsumerImpl.fs [hunk 2]
```diff
@@ -32,26 +33,26 @@ type internal ConsumerMessage<'T> =
     | ConnectionClosed of ClientCnx
     | ReachedEndOfTheTopic
     | MessageReceived of RawMessage * ClientCnx
-    | Receive of CancellationToken * AsyncReplyChannel<ResultOrException<Message<'T>>>
-    | BatchReceive of CancellationToken * AsyncReplyChannel<ResultOrException<Messages<'T>>>
+    | Receive of CancellationToken * TaskCompletionSource<ResultOrException<Message<'T>>>
+    | BatchReceive of CancellationToken * TaskCompletionSource<ResultOrException<Messages<'T>>>
     | SendBatchByTimeout
-    | Acknowledge of MessageId * AckType * Transaction option * AsyncReplyChannel<Task<Unit>> option
+    | Acknowledge of MessageId * AckType * Transaction option * TaskCompletionSource<Task<Unit>> option
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConsumerImpl.fs:39]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/ConsumerImpl.fs [hunk 3]
```diff
@@ -32,26 +33,26 @@ type internal ConsumerMessage<'T> =
...
     | ActiveConsumerChanged of bool
-    | Close of AsyncReplyChannel<ResultOrException<unit>>
-    | Unsubscribe of AsyncReplyChannel<ResultOrException<unit>>
+    | Close of TaskCompletionSource<ResultOrException<unit>>
+    | Unsubscribe of TaskCompletionSource<ResultOrException<unit>>
     | Tick of ConsumerTickType
-    | GetStats of AsyncReplyChannel<ConsumerStats>
-    | ReconsumeLater of Message<'T> * AckType * TimeStamp * AsyncReplyChannel<Task<unit>>
+    | GetStats of TaskCompletionSource<ConsumerStats>
+    | ReconsumeLater of Message<'T> * AckType * TimeStamp * TaskCompletionSource<Task<unit>>
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/ConsumerImpl.fs:50]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs [hunk 4]
```diff
@@ -39,29 +40,29 @@ type internal BatchAddResponse<'T> =
...
-    | AcknowledgeCumulative of AsyncReplyChannel<Task<unit>> * MessageId * Transaction option
-    | RedeliverUnacknowledged of RedeliverSet * AsyncReplyChannel<Task>
-    | RedeliverAllUnacknowledged of AsyncReplyChannel<Task>
-    | Close of AsyncReplyChannel<ResultOrException<unit>>
-    | Unsubscribe of AsyncReplyChannel<ResultOrException<unit>>
-    | HasReachedEndOfTheTopic of AsyncReplyChannel<bool>
-    | Seek of SeekType * AsyncReplyChannel<Task>
-    | SeekWithResolver of Func<string, SeekType> *  AsyncReplyChannel<Task> 
+    | Acknowledge of TaskCompletionSource<Task<unit>> * MessageId * Transaction option
+    | NegativeAcknowledge of TaskCompletionSource<Task<unit>> * MessageId
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:48]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs [hunk 5]
```diff
@@ -39,29 +40,29 @@ type internal BatchAddResponse<'T> =
...
+    | AcknowledgeCumulative of TaskCompletionSource<Task<unit>> * MessageId * Transaction option
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:49]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs [hunk 6]
```diff
@@ -39,29 +40,29 @@ type internal BatchAddResponse<'T> =
...
-    | NegativeAcknowledge of AsyncReplyChannel<Task<unit>> * MessageId
-    | AcknowledgeCumulative of AsyncReplyChannel<Task<unit>> * MessageId * Transaction option
-    | RedeliverUnacknowledged of RedeliverSet * AsyncReplyChannel<Task>
-    | RedeliverAllUnacknowledged of AsyncReplyChannel<Task>
-    | Close of AsyncReplyChannel<ResultOrException<unit>>
-    | Unsubscribe of AsyncReplyChannel<ResultOrException<unit>>
-    | HasReachedEndOfTheTopic of AsyncReplyChannel<bool>
-    | Seek of SeekType * AsyncReplyChannel<Task>
-    | SeekWithResolver of Func<string, SeekType> *  AsyncReplyChannel<Task> 
+    | Acknowledge of TaskCompletionSource<Task<unit>> * MessageId * Transaction option
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:47]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs [hunk 7]
```diff
@@ -39,29 +40,29 @@ type internal BatchAddResponse<'T> =
...
+    | NegativeAcknowledge of TaskCompletionSource<Task<unit>> * MessageId
+    | AcknowledgeCumulative of TaskCompletionSource<Task<unit>> * MessageId * Transaction option
+    | RedeliverUnacknowledged of RedeliverSet * TaskCompletionSource<Task>
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:50]: It's essentially `Task<Task>`, can we leave just one?

File: src/Pulsar.Client/Internal/TaskSeq.fs [hunk 8]
```diff
@@ -6,12 +6,13 @@ open System.Threading.Tasks
...
 open Pulsar.Client.Common
 open Microsoft.Extensions.Logging
+open System.Threading.Channels
 
 type internal TaskGenerator<'T> = unit -> Task<'T>
 
 type internal TaskSeqMessage<'T> =
-    | WhenAnyTask of AsyncReplyChannel<Task<Task<'T>>>
-    | Next of AsyncReplyChannel<Task<Task<'T>>>
+    | WhenAnyTask of TaskCompletionSource<Task<Task<'T>>>
```
Comment (Lanayx) [inline src/Pulsar.Client/Internal/TaskSeq.fs:14]: It's essentially `Task<Task>`, can we leave just one?

**Review by Lanayx — CHANGES_REQUESTED** [review]

File: src/Pulsar.Client/Internal/ConsumerImpl.fs (hunk 1 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/ConsumerImpl.fs]: Done.

File: src/Pulsar.Client/Internal/ConsumerImpl.fs (hunk 2 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/ConsumerImpl.fs:39]: This one couldn't be done easily.

File: src/Pulsar.Client/Internal/ConsumerImpl.fs (hunk 3 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/ConsumerImpl.fs:50]: I haven't had a chance to look at it and I suggest to look at it in the separate PR

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs (hunk 5 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:49]: I haven't had a chance to look at it and I suggest to look at it in the separate PR

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs (hunk 6 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:47]: I haven't had a chance to look at it and I suggest to look at it in the separate PR

File: src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs (hunk 7 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/MultiTopicsConsumerImpl.fs:50]: I haven't had a chance to look at it and I suggest to look at it in the separate PR

File: src/Pulsar.Client/Internal/TaskSeq.fs (hunk 8 above)
Comment (TiraelSedai) [reply src/Pulsar.Client/Internal/TaskSeq.fs:14]: I haven't had a chance to look at it and I suggest to look at it in the separate PR

**Pull request merged**